File: Java/Token.java
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Token {

    private final String key;
    private Object[] data;

    public Token(String key) {
        this.key = key;
    }

    public void setData(Object... data) {
        this.data = data;
    }

    public String generate(
            long time
            )
            throws
            NoSuchAlgorithmException,
            InvalidKeyException,
            IOException
            {
        Mac mac = Mac.getInstance("HmacSHA1");
        mac.init(new SecretKeySpec(
                key.getBytes(), "RAW")
                );
        StringBuilder buffer = new StringBuilder();
        if(data != null) {
            for(Object object : data) {
                buffer.append(object);
            }
        }
        byte[] array = buffer.toString().getBytes("UTF-8");
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] digest = md.digest(
                mac.doFinal(
                        ByteBuffer.allocate(array.length + 8)
                                  .put(array)
                                  .putLong(time)
                                  .array()
                        )
                );
        return new String(new sun.misc.BASE64Encoder().encode(digest));
    }
}

File: C#/Token.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Linq;
using System.IO;
using System.Text;

namespace STREAM
{
    class Token
    {
        private String key;
        private Object[] data;

        public Token(String key)
        {
            this.key = key;
        }

        public void SetData(params Object[] data)
        {
            this.data = data;
        }

        public String Generate(long time)
        {
            StringBuilder builder = new StringBuilder();
            if (data != null)
            {
                foreach (Object o in data)
                {
                    builder.Append(o);
                }
            }
            byte[] array = Encoding.UTF8.GetBytes(
                builder.ToString()
                ).Concat(
                BitConverter.GetBytes(time).Reverse().ToArray()
                ).ToArray();

            HMACSHA1 mac = new HMACSHA1(Encoding.ASCII.GetBytes(key));
            return Convert.ToBase64String(
                MD5.Create().ComputeHash(
                    mac.ComputeHash(array)
                    )
                    );
        }
    }
}